[TDI-66] Make dialogs in page editor preview local

sebxian opened the pull request:
## Description

Adds ability to create dialogs that don't blur out the whole screen but a local are, and add them to the page editor so in-page dialogs will only blur out the page preview (and not the editor).

By the way, this is also used in the page's live preview:

![image](https://github.com/user-attachments/assets/f2d97857-ec02-41c7-8fc2-fff6f801d180)

@stefan-kracht , improvements are possible, but I am not sure about the approach to pursue so if you have time, I'd be grateful for your two cents at this stage!

Specifically:
* I updated the Design component to accept a "container" prop and handle this use case correctly. I find the part with the conditional styling (absolute vs fixed) not elegant. On the other hand, it just works, and no boilerplate needs to be added to create an alternative Dialog component.
* I've added a `<LocalDialogContainer />`, which is a context, exporting `useLocalDialog()` to get the ref to the container from a descendant in the DOM. This is quite easy to use. I've also thought about adding a `LocalDialogContent` that hides away importing the hook... but it would just wrap the design Dialog at this point and I'm not sure it's worth it.
* Unless I am missing something, just using the regular composition pattern for dialogs does not work, because the dialog definitions can be nested somewhere deep in the DOM and a context is needed to "find" the container.

Let me know what you think, I am happy to revisit this tomorrow and see if I can find a way to improve it further.

## Checklist

- [x] Documentation updated if required
- [x] Test coverage is appropriate
      
## Checklist Internal

- [x] Linear issue linked (e.g. [DIR-XXXX] pull request title)
- [x] Has the PR been labeled

stefan-kracht reviewed (changes requested):
Very nice. This works great. 

> I updated the Design component to accept a "container" prop and handle this use case correctly. I find the part with the conditional styling (absolute vs fixed) not elegant. On the other hand, it just works, and no boilerplate needs to be added to create an alternative Dialog component.

I actually like that approach. It is the minimal required CSS change to adjust that behavior. I like it much better than adding an alternative container component.

> Unless I am missing something, just using the regular composition pattern for dialogs does not work, because the dialog definitions can be nested somewhere deep in the DOM and a context is needed to "find" the container.

Not sure if I understand that. You did NOT change the composition pattern, did you? We still place the same components in the same order, right?

One problem we have in the page editor is that, when you add content to a dialog and reach for the sidebar form, you close the dialog (I also noticed that the dialog adds `pointer-events: none;` to the body tag. Removing that in the dev tools does not solve it, but maybe it is a hint at least. I would expected that to be added to the relative container and not the body in our new)

Some small things:
- I think it would be nice to add a story to storybook to document that new feature
- I also added one comment in the code

sebxian commented on ui/src/components/LocalDialog/index.tsx:
```diff
@@ -0,0 +1,38 @@
+  children,
+  className,
+}: LocalDialogContainerProps) => {
+  const [container, setContainer] = useState<HTMLDivElement | null>(null);
```
> Not sure if I understand that. You did NOT change the composition pattern, did you? We still place the same components in the same order, right?

Correct. My initial idea was that it could be somehow be

```
<LocalDialog>
  < other stuff />
  <DialogTrigger />
  <DialogContent />
</LocalDialog>

And these nested components would just automagically sort out the ref.

stefan-kracht replied:
Okay, this pattern was new to me. I did a little research and testing with both versions and this is how I understand it:

TLDR: The code from the Radix docs is the only correct solution here

- `useRef` will avoid a rerendering of that component and `useState` will not  
-  When the `LocalDialogContainer` mounts the `div` is not there just yet (it exists in JSX, but the corresponding DOM element is not yet created)  
-  But the `ref` attribute of the `div` element will be called after the component mounts (this is just what the `ref` is for. To announce when it was mounted)  
-  When this ref function is a `useState` function it will rerender the component after being called and if it is a `MutableRefObject` from a `useRef` it will not (that's the point of a `useRef`)  
-  My suggestion with a `useRef` works but only by accident because something else is triggering a rerender of that part of the tree (when the page loads the YAML file, it triggers a rerender of that tree)

sebxian commented:
@stefan-kracht I have made quite a few changes because it was no simple task to get an overlay that only blocks interaction with part of the page (this doesn't seem to be supported by radix). 

So in a nutshell:
* LocalDialog now is a separate component that doesn't rely on the design component.
  * I couldn't use radix' overlay but had to roll my own, and the composition isn't exactly the same. As a result, LocalDialog draws on some radix primitives and uses similar styles as the design component (these were copied over and adapted).
  * I removed the "container" prop from the design component because I ended up not using it.
* LocalDialog now calls the hook `useLocalDialogContainer()` to find the ref, so no prop needs to be defined and it "just works" as long as the components are properly nested.
* There is also an example of this in Storybook.
* I needed to tweak the mechanic for focusing a block (that is, loading the side panel editor) to make sense with Dialog blocks. Please test this and give feedback to make sure the behavior makes sense.
* I removed the prop "showCloseButton" from the design DialogContent component and instead exported a <DialogXClose /> component. This way, I could also import it to the LocalDialog component and we're keeping the compositional pattern used with other Dialog sub components. I hope you agree with this change; to me it feels a bit better but that is probably a matter of taste.

Things that don't feel final yet but we should probably circle back to them when the look and feel of the whole page editor is closer to a finished product:
* The exact positioning of the local overlay. I gave it a negative margin because I wanted it a bit wider than the elements "hidden" under it. It may also be nice to add rounded corners? But it's probably too early for fine-tuning at this point.
* Adding blocks inside of dialogs still feels a bit clunky, but that should be revisited when the drag and drop method has been added.

stefan-kracht reviewed (approved):
Very nice. I like how you also covered the UI problem of what should be selected in the side panel when opening or closing the dialog. And I appreciate the refactoring of the return value of `findAncestor`. It even makes the test much more robust against false positives. I was changing the complex examples the other day and had a lot of failing tests with `findAncestor` and fixing these would now be much easier.

Regarding having rounded corners for the backdrop: I think this is fine as it is right now. Since we also use the local dialog on the live page, rounded corners would look odd in that context.

I only have one suggestion: Can we move the `LocalDialog` into the design folder as well? I think it does not have any application-specific dependencies, and moving it into the design folder would make this even clearer.

Approved anyway...
